tests/php/ORM/HierarchyCachingTest.php

```diff
@@ -0,0 +1,150 @@
+            $actual = $node->numChildren($cache);
+            $this->assertEquals($expected, $actual, $message);
+        }
+    }
+
+    public function prepopulateCacheNumChildrenDataProvider()
+    {
+        return [
+            [
+                TestObject::class, 'Stage', [],
+                TestObject::class, 'obj1', false, 0, 'childless object should have a numChildren of 0'
+            ],
+            [
+                TestObject::class, 'Stage', [],
+                TestObject::class, 'obj1', true, 0, 'childless object should have a numChildren of 0 when cache'
+            ],
+            [
+                TestObject::class, 'Stage', [2],
+                TestObject::class, 'obj1', false, 0, 'childless object should have a numChildren of 0'
+            ],
+            [
+                TestObject::class, 'Stage', [2],
+                TestObject::class, 'obj1', true, 0, 'childless object should have a numChildren of 0 when cache'
+            ],
+            [
+                TestObject::class, 'Stage', [],
+                TestObject::class, 'obj2', false, 2, 'Root object numChildren should count direct children'
+            ],
+            [
+                TestObject::class, 'Stage', [],
+                TestObject::class, 'obj2', true, 2, 'Root object numChildren should count direct children when cache'
+            ],
+            [
+                TestObject::class, 'Stage', [2],
+                TestObject::class, 'obj2', false, 2, 'Root object numChildren should count direct children'
+            ],
+            [
+                TestObject::class, 'Stage', [2],
+                TestObject::class, 'obj2', true, 2, 'Root object numChildren should count direct children when cache'
+            ],
+            [
+                HideTestObject::class, 'Stage', [],
+                HideTestObject::class, 'obj4', false, 1, 'Hidden object should not be included in count'
+            ],
+            [
+                HideTestObject::class, 'Stage', [],
+                HideTestObject::class, 'obj4', true, 1, 'Hidden object should not be included in count when cache'
+            ],
+            [
+                HideTestObject::class, 'Stage', [2],
+                HideTestObject::class, 'obj4', false, 1, 'Hidden object should not be included in count'
+            ],
+            [
+                HideTestObject::class, 'Stage', [2],
+                HideTestObject::class, 'obj4', true, 1, 'Hidden object should not be included in count when cache'
+            ]
+        ];
+    }
+
+    /**
+     * @dataProvider prepopulateCacheNumChildrenDataProvider
+     */
+    public function testPrepopulatedNumChildrenCache(
+        $baseClass,
+        $stage,
+        $idList,
+        $className,
+        $identifier,
+        $cache,
+        $expected,
+        $message
+    ) {
+        Hierarchy::prepopulate_numchildren_cache($baseClass, $stage, $idList);
```

Comment:
We may want to test this instead, due to my most recent addition:

```
TestObject::singleton()->prepopulateTreeDataCache($idLIst, [ 'numChildrenMethod' => 'numChildren' ]);
```

Compare to how CMS is calling things in the relevant PR here: https://github.com/silverstripe/silverstripe-cms/pull/2266/files

Reply:
Done.